### PR TITLE
[Filtering] Add advanced rules filtering to MySQL and extend the BaseDataSource API

### PR DESCRIPTION
**Pull request by timgrein**
## Closes https://github.com/elastic/enterprise-search-team/issues/3596 ##

This PR introduces advanced rules filtering for MySQL by routing custom SQL queries to the specified database/table pairs. 
Also extended the API for the `BaseDataSource` so that `get_docs` can accept parameters like filtering. Added two convenient methods: one to check whether advanced rules are present and one to extract them (so changes on the protocol side don't affect concrete implementations, but only the framework).

Further refactorings like DRY out `test_mysql` with the newly introduced `setup_mysql_source` method, improvements around docs/variable names and speed up of one existing test for connection retrying will happen in separate PRs as this PR is focused on the functionality.

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
~~- [ ] Considered corresponding documentation changes~~ (will follow with a separate PR)
~~- [ ] Contributed any configuration settings changes to the configuration reference~~

## Release Note

Add advanced rules filtering capabilities to the framework and the MySQL connector.

### CODE REVIEW

File: connectors/sources/tests/test_mysql.py [hunk 1]
```diff
@@ -188,6 +246,7 @@ async def test_ping_negative(patch_logger):
             await source.ping()
 
 
+@pytest.mark.fail_slow(10)
```
Comment (timgrein) [inline connectors/sources/tests/test_mysql.py:249]: This is intentional and will be fixed with a separate PR. Test was always this slow, but shouldn't be the focus in this PR.

Comment (timgrein) [inline connectors/sources/tests/test_mysql.py:249]: Will be fixed with this [PR](https://github.com/elastic/connectors-python/pull/379)

**Review by tarekziade — COMMENTED** [review]
looks good, just one proposal to structure filtering, and the get_docs signature

File: connectors/source.py [hunk 2]
```diff
@@ -104,6 +104,22 @@ def is_empty(self):
         return len(self._config) == 0
 
 
+def advanced_rules_present(filtering=None):
```
[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/source.py:107]: We're not passing the Filtering object, but the dictionary directly. Thought about it, but would complicate some stuff IMO.

File: connectors/sources/abs.py [hunk 3]
```diff
@@ -250,7 +250,7 @@ async def get_blob(self, container):
                     blob=blob, container_metadata=container["metadata"]
                 )
 
-    async def get_docs(self):
+    async def get_docs(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/sources/abs.py:253]: Also had it like this way before. I thought it would be better for connectors which don't need filtering to keep it this way. But happy to change it to have it more explicit.

File: connectors/source.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarekziade) [reply connectors/source.py:107]: Fair enough. it's just that I identify those two functions as being properties of that dict. So maybe that dict will need to become a class with more behavior. Maybe they need to move to `connectors.filtering.*`

File: connectors/sources/abs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tarekziade) [reply connectors/sources/abs.py]: I am thinking about the base class signature and documenting the behavior see https://github.com/elastic/connectors-python/blob/main/connectors/source.py#L194-L219

if calling `get_docs` has just that one optional argument and nothing else. I would set it as an explicit arg.

[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/sources/abs.py]: Adjusted it

**Review by tarekziade — APPROVED** [review]

File: connectors/byoc.py [hunk 4]
```diff
@@ -268,14 +271,34 @@ def get_draft_filter(self, domain=DEFAULT_DOMAIN):
...
-                filter_[filter_state]
+                Filter(filter_[filter_state])
                 for filter_ in self.filtering
                 if filter_["domain"] == domain
             ),
             {},
         )
 
 
+class Filter(dict):
```
Comment (tarekziade) [inline connectors/byoc.py:282]: there's a cleaner way to extend a dict by the way, see https://stackoverflow.com/a/3387975/15097096

**Comment by timgrein** [comment]
buildkite test this
